// elements-zone-strategy/bundles/elements-zone-strategy.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/elements'), require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('elements-zone-strategy', ['exports', '@angular/elements', '@angular/core'], factory) :
    (global = global || self, factory(global['elements-zone-strategy'] = {}, global.ng.elements, global.ng.core));
}(this, (function (exports, elements, core) { 'use strict';

    var ElementZoneStrategy = /** @class */ (function () {
        function ElementZoneStrategy(strategy, injector) {
            this.strategy = strategy;
            this.injector = injector;
            this.ngZone = this.injector.get(core.NgZone);
            if (this.strategy.events) {
                this.events = this.strategy.events;
            }
        }
        ElementZoneStrategy.prototype.connect = function (element) {
            var _this = this;
            this.runInZone(function () { _this.strategy.connect(element); });
            if (!this.events) {
                this.events = this.strategy.events;
            }
        };
        ElementZoneStrategy.prototype.disconnect = function () {
            var _this = this;
            this.runInZone(function () { _this.strategy.disconnect(); });
        };
        ElementZoneStrategy.prototype.getInputValue = function (propName) {
            var _this = this;
            // tslint:disable-next-line:arrow-return-shorthand
            return this.runInZone(function () { return _this.strategy.getInputValue(propName); });
        };
        ElementZoneStrategy.prototype.setInputValue = function (propName, value) {
            var _this = this;
            this.runInZone(function () { _this.strategy.setInputValue(propName, value); });
        };
        ElementZoneStrategy.prototype.runInZone = function (fn) { return this.ngZone.run(fn); };
        return ElementZoneStrategy;
    }());

    var ElementZoneStrategyFactory = /** @class */ (function () {
        function ElementZoneStrategyFactory(component, injector) {
            this.component = component;
            this.injector = injector;
            this.ngElement = elements.createCustomElement(this.component, { injector: this.injector });
            // this is to keep the document-register-element polyfill happy
            // it doesn't like creating elements before the define, so we fill the registry with a random definition
            customElements.define(this.generateName(), this.ngElement);
        }
        ElementZoneStrategyFactory.prototype.string4 = function () {
            return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
        };
        ElementZoneStrategyFactory.prototype.generateName = function () {
            var result = 'dummy-name-n';
            var temp = ElementZoneStrategyFactory.counter + '';
            result = result + temp + '-' + this.string4() + this.string4() + this.string4() + '-' +
                this.string4() + this.string4() + '-' + this.string4();
            ElementZoneStrategyFactory.counter++;
            return result;
        };
        ElementZoneStrategyFactory.prototype.create = function (injector) {
            // the only way to get a default strategy outside @angular/elements
            // is to create the ngElement/ngElementImpl and get it from a property
            var tempElement = new this.ngElement(this.injector);
            var strategy = tempElement['ngElementStrategy'];
            tempElement = null;
            return new ElementZoneStrategy(strategy, this.injector);
        };
        return ElementZoneStrategyFactory;
    }());
    ElementZoneStrategyFactory.counter = 1;

    /*
     * Public API Surface of elements-zone-strategy
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ElementZoneStrategyFactory = ElementZoneStrategyFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
