// elements-zone-strategy/fesm2015/elements-zone-strategy.js
import { createCustomElement } from '@angular/elements';
import { NgZone } from '@angular/core';

class ElementZoneStrategy {
    constructor(strategy, injector) {
        this.strategy = strategy;
        this.injector = injector;
        this.ngZone = this.injector.get(NgZone);
        if (this.strategy.events) {
            this.events = this.strategy.events;
        }
    }
    connect(element) {
        this.runInZone(() => { this.strategy.connect(element); });
        if (!this.events) {
            this.events = this.strategy.events;
        }
    }
    disconnect() {
        this.runInZone(() => { this.strategy.disconnect(); });
    }
    getInputValue(propName) {
        // tslint:disable-next-line:arrow-return-shorthand
        return this.runInZone(() => { return this.strategy.getInputValue(propName); });
    }
    setInputValue(propName, value) {
        this.runInZone(() => { this.strategy.setInputValue(propName, value); });
    }
    runInZone(fn) { return this.ngZone.run(fn); }
}

class ElementZoneStrategyFactory {
    constructor(component, injector) {
        this.component = component;
        this.injector = injector;
        this.ngElement = createCustomElement(this.component, { injector: this.injector });
        // this is to keep the document-register-element polyfill happy
        // it doesn't like creating elements before the define, so we fill the registry with a random definition
        customElements.define(this.generateName(), this.ngElement);
    }
    string4() {
        return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
    }
    generateName() {
        let result = 'dummy-name-n';
        const temp = ElementZoneStrategyFactory.counter + '';
        result = result + temp + '-' + this.string4() + this.string4() + this.string4() + '-' +
            this.string4() + this.string4() + '-' + this.string4();
        ElementZoneStrategyFactory.counter++;
        return result;
    }
    create(injector) {
        // the only way to get a default strategy outside @angular/elements
        // is to create the ngElement/ngElementImpl and get it from a property
        let tempElement = new this.ngElement(this.injector);
        const strategy = tempElement['ngElementStrategy'];
        tempElement = null;
        return new ElementZoneStrategy(strategy, this.injector);
    }
}
ElementZoneStrategyFactory.counter = 1;

/*
 * Public API Surface of elements-zone-strategy
 */

/**
 * Generated bundle index. Do not edit.
 */

export { ElementZoneStrategyFactory };
